### PR TITLE
Feature/tester qa specs

### LINKED ISSUE #173: Parametros para el rol de tester
[X] Refactoriza tests existentes en js/test/script.spec.js para nueva arquitectura 
 [x] Crea js/test/models.spec.js con suite completa de tests para clases POO: 
 [x] Crea js/test/storage.spec.js con tests de operaciones CRUD y manejo de errores 
 [x] Actualiza js/test/testing-doc.md con una nueva estructura de testing y métricas de cobertura actualizadas 
 [x] Reporta issues detalladas en GitHub para bugs encontrados.

### PR DESCRIPTION
**Pull request by Santi22-7**
# 🟣 PULL REQUEST – Actividad 4 – Programación Web I

---

## 📌 Datos del Estudiante

- **Nombre y Apellido: Santiago Ariel Samitier**  
- **Número de Matrícula: 148249**  
- **Carrera:** Tecnicatura Universitaria en Programación de Sistemas  
- **Materia:** Programación Web I  
- **Profesor:** Lic. Matías Velasquez  
- **Cuatrimestre/Año:** 1º Cuatrimestre / 2026
- **Rol asignado para esta entrega:  Coordinador / DevOps + Tester QA ** 

---

## 📂 Rama de trabajo

- **Nombre de la rama:** `feature/tester-qa-specs`
 
---

## 📝 Descripción del trabajo realizado

Se realizo los test de lo echo en esta actividad, Se realizo la documentación correspondiente 

---

## 📄 Archivos modificados / agregados

- `Changelog.md`
- `test-runner.html`   
- `spec-tester-qa.md`
- `models.spec.js`
- `storage.spec.js`
- `script.spec.js`  
- `Test-pass-Actividad-4.png`
- `Suite-1.png`
- `Suite-2.png`
- `Suite-3.png`

---
Issues cerradas:
Closes #173 
---
##  Cheklist:
 [X] Refactoriza tests existentes en js/test/script.spec.js para nueva arquitectura 
 [x] Crea js/test/models.spec.js con suite completa de tests para clases POO: 
 [x] Crea js/test/storage.spec.js con tests de operaciones CRUD y manejo de errores 
 [x] Actualiza js/test/testing-doc.md con una nueva estructura de testing y métricas de cobertura actualizadas 
 [x] Reporta issues detalladas en GitHub para bugs encontrados.

### CODE REVIEW

**Comment by hmarc953** [comment]
# CODE REVIEW - PR #174

---

## 🔍 HALLAZGOS DETALLADOS

### HALLAZGO #1: Falta sección "AT CLOSE" en spec-tester-qa.md

**Archivo:** `docs/03-specs/actividad-obligatoria-4/spec-tester-qa.md`  
**Línea:** 1-638  
**Tipo de problema:** Legibilidad / Documentación Incompleta  
**Severidad:** MEDIA

#### Explicación técnica
El archivo `spec-tester-qa.md` contiene únicamente la sección "BEFORE" (plan de testing, herramientas, criterios de aceptación). Según la consigna, debe incluir obligatoriamente una sección "AT CLOSE" que documente:
- Prompt exacto utilizado en Copilot Agent para generar models.spec.js y storage.spec.js
- Screenshots del test runner ejecutado con Playwright MCP
- Resumen: cuántos tests pasaron, cuántos fallaron, cuántos bugs se reportaron

La falta de esta sección impide verificar que se utilizó el flujo obligatorio de documentación (Prompt → Output → Ajustes).

#### Sugerencia de mejora
Completar `spec-tester-qa.md` con sección "AL CERRAR LA TAREA":

---

### HALLAZGO #2: Cobertura incompleta de tests en models.spec.js

**Archivo:** `js/test/models.spec.js`  
**Línea:** 1-281  
**Tipo de problema:** Cobertura de Tests Incompleta  
**Severidad:** MEDIA

#### Explicación técnica
El archivo models.spec.js contiene tests para clases POO pero hay inconsistencias:

1. Se importan todas las 7 clases (Usuario, Pelicula, Funcion, Compra, CatalogoPeliculas, GestorUsuarios, ConsultaSoporte)
2. Solo se testea `Usuario` en la sección visible
3. Faltan suites para: Pelicula, Funcion, Compra, CatalogoPeliculas, GestorUsuarios, ConsultaSoporte
4. El archivo parece truncado en la presentación del diff

La consigna requiere testing de "7 clases y métodos de POO" de forma completa.

#### Sugerencia de mejora
Completar el archivo models.spec.js para incluir tests de todas las 7 clases:

```javascript
describe('Pelicula', function() {
  it('crea película con valores normalizados', function() {
    const pelicula = new Pelicula('p1', 'Avatar', 'CF', 2009, 8.5, 'Acción', ['Director', 'Actor']);
    expect(pelicula.id).toBe('p1');
    expect(pelicula.titulo).toBe('Avatar');
    // ... más tests
  });
});

describe('Funcion', function() {
  it('calcula disponibilidad correctamente', function() {
    const funcion = new Funcion('f1', 'CineMax', 'Español', '20:00', 30, 15.5);
    expect(funcion.hayDisponibilidad(5)).toBe(true);
    // ... más tests
  });
});

describe('Compra', function() {
  it('calcula total con descuentos', function() {
    // ... tests de lógica de compra
  });
});

// ... Suites para CatalogoPeliculas, GestorUsuarios, ConsultaSoporte
```

---

### HALLAZGO #3: script.spec.js refactorizado sin cobertura clara de nueva arquitectura

**Archivo:** `js/test/script.spec.js`  
**Línea:** Refactorización completa (206 líneas eliminadas)  
**Tipo de problema:** Cobertura de Tests Incompleta  
**Severidad:** BAJA

#### Explicación técnica
El archivo script.spec.js fue refactorizado eliminando 206 líneas de tests antiguos (que validaban funciones con prompt/alert). Sin embargo, la nueva versión no muestra tests claros que validen:

1. Inicialización correcta del controlador (DOMContentLoaded)
2. Event listeners registrados correctamente en elementos del DOM
3. Interacción correcta entre controlador y clases POO del dominio
4. Manipulación del DOM en respuesta a eventos

El diff solo muestra el cambio en test-runner.html pero no los tests nuevos en script.spec.js.

#### Sugerencia de mejora
Agregar tests en script.spec.js que cubran la nueva arquitectura:

```javascript
describe('Controlador - Eventos y DOM', function() {
  it('registra event listeners al cargar la página', function() {
    // Verificar que event listeners estén registrados en botones
    const botonCompra = document.querySelector('#btn-compra');
    expect(botonCompra).toBeTruthy();
  });

  it('actualiza DOM cuando se crea un usuario', function() {
    const usuario = new Usuario('u1', 'Juan', 'juan@test.com', 'Pass123');
    // Simular acción en controlador
    // Verificar que DOM se actualizó
    expect(document.querySelector('.usuario-nombre')).toBeTruthy();
  });

  it('valida entrada antes de invocar métodos POO', function() {
    // Test de validación en el controlador
  });
});
```

---

### HALLAZGO #4: Falta ejecución con Playwright MCP

**Archivo:** Evidencia de ejecución  
**Línea:** N/A  
**Tipo de problema:** Falta de Evidencia  
**Severidad:** MEDIA

#### Explicación técnica
La consigna especifica explícitamente que el Tester QA **DEBE**:
> "Utilizar Playwright MCP (@playwright/mcp) para abrir `js/test/test-runner.html` en un browser real, ejecutar las suites de Jasmine y capturar screenshots de los resultados (PASS/FAIL)"

**No hay evidencia en el PR** de que se haya ejecutado Playwright MCP:
- No hay transcripción de comandos Playwright MCP
- El test-runner.html no fue ejecutado en browser real documentado
- Los screenshots mencionados ("Test-pass-Actividad-4.png", "Suite-1.png", etc.) no están embebidos en testing-doc.md

#### Sugerencia de mejora
Ejecutar con Playwright MCP:

```bash
playwright open test-runner.html --screenshot
```

Capturar screenshots mostrando:
1. Suite 1 (Usuario): indicador de PASS/FAIL count
2. Suite 2 (Storage): indicador de PASS/FAIL count
3. Suite 3 (Eventos/DOM): indicador de PASS/FAIL count
4. Resumen final de cobertura

Incrustar en `testing-doc.md`:
```markdown
### Ejecución de Tests con Playwright MCP

![Suite 1 - Usuario Tests](../test-results/suite-1-usuario.png)
![Suite 2 - Storage Tests](../test-results/suite-2-storage.png)
![Suite 3 - DOM Tests](../test-results/suite-3-dom.png)
```

---

### HALLAZGO #5: testing-doc.md sin métricas de cobertura cuantificadas

**Archivo:** `js/test/testing-doc.md`  
**Línea:** 247-331 (sección agregada)  
**Tipo de problema:** Documentación Incompleta  
**Severidad:** BAJA

#### Explicación técnica
La actualización a testing-doc.md incluye secciones de "Suites de Tests" pero:

1. **Métricas no cuantificadas:** No dice "X% de cobertura de líneas" ni "Y% de cobertura de funciones"
2. **Sección Capturas de Pantalla vacía:** Aparece pero sin images embebidas
3. **Issues Conocidos vacía:** Lista vacía o ausente

#### Sugerencia de mejora
Actualizar testing-doc.md con:

```markdown
## Métricas de Cobertura

- **Cobertura de Líneas:** 82%
- **Cobertura de Funciones:** 88%
- **Cobertura de Ramas:** 75%

| Suite | Tests | PASS | FAIL | Coverage |
|-------|-------|------|------|----------|
| Usuario | 15 | 15 | 0 | 95% |
| Pelicula | 12 | 12 | 0 | 90% |
| Funcion | 10 | 10 | 0 | 85% |
| Compra | 14 | 14 | 0 | 92% |
| CatalogoPeliculas | 8 | 8 | 0 | 80% |
| GestorUsuarios | 9 | 9 | 0 | 88% |
| ConsultaSoporte | 6 | 6 | 0 | 75% |
| Storage | 18 | 18 | 0 | 85% |
| DOM/Eventos | 5 | 5 | 0 | 70% |
| **TOTAL** | **97** | **97** | **0** | **85%** |

## Capturas de Pantalla

### Suite 1: Usuario Tests
[Captura mostrando 15 tests PASS]

### Suite 2: Storage Tests
[Captura mostrando 18 tests PASS]

### Suite 3: DOM/Eventos Tests
[Captura mostrando 5 tests PASS]

## Issues Conocidos

- [#175](https://github.com/hmarc953/cineglobal/issues/175) - Bug en validación de email
- [#176](https://github.com/hmarc953/cineglobal/issues/176) - Edge case en manejo de disponibilidad
```

---

## 🎯 DECISION FINAL

### **REQUEST CHANGES**

**Razones principales:**

1. **Documentación obligatoria incompleta:** Falta sección "AT CLOSE" en spec-tester-qa.md que es requisito explícito de la consigna
2. **Ejecución no verificable:** No hay evidencia de ejecución con Playwright MCP (requisito obligatorio)
3. **Cobertura de tests incompleta:** Solo ~2 de 7 clases testeadas (visible en diff)
4. **Métricas no cuantificadas:** Falta información concreta de cobertura de tests

**Comment by Santi22-7** [comment]
La corrección 1 esta mal porque el   "AT CLOSE" esta en el código del archivo

File: docs/03-specs/actividad-obligatoria-4/spec-tester-qa.md [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+**Versión**: 1.0  
+**Última actualización**: 2026-06-08
+
+## AT CLOSE:
+
+### Prompt :
+"bien en el archivo script.spec.js solo debe estar los test para eventos y dom. Y en el archivo models.spec.js deben estar los test solo para los modelos .Y para storage.spec.js solo los test para storage . Si tenes que crear nuevos test utiliza el archivo spec-tester-qa.md como contexto"
+
+### Screenshots del test runner:
+- Screenshots: [Screenshots](//js/test/screenshots/Test-pass-Actividad-4.png)
```
Comment (hmarc953) [inline docs/03-specs/actividad-obligatoria-4/spec-tester-qa.md]: Path esta mal. Corregir a:
`/js/test/screenshots/Test-pass-Actividad-4.png`

**Review by hmarc953 — CHANGES_REQUESTED** [review]

**Review by hmarc953 — APPROVED** [review]
